### PR TITLE
Warn on duplicate characters

### LINKED ISSUE #3: Remove duplicate symbols from alphabet
https://twitter.com/kelin2025/status/961573925489700864

### PR DESCRIPTION
**Pull request by Kelin2025**
Fixed #3  
P.S. Sorry for code format. It would be better to have some .eslintrc/.prettierrc in repository for these situations

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,114 +1,113 @@
...
-calculate();
-
-function calculate () {
+function calculate() {
   var input = {
     size: Number(sizeInput.value),
     alphabet: alphabetInput.value,
+    uniqAlphabet: alphabetInput.value.split('').reduce(function(res, letter) {
+      return res.includes(letter) ? res : res + letter
+    }, ''),
```
Comment (Kelin2025) [inline src/index.js]: Added uniqAlphabet field here

File: src/index.js [hunk 2]
```diff
@@ -1,114 +1,113 @@
...
 
-function result (input) {
-  resultWrapper.style.display = 'none';
+function result(input) {
+  resultWrapper.style.display = 'none'
 
-  if (input.alphabet.length < 2) {
-    return;
+  if (input.uniqAlphabet.length < 2) {
+    return
```
Comment (Kelin2025) [inline src/index.js]: Replaced `alphabet` to `uniqAlphabet`

File: src/index.js [hunk 3]
```diff
@@ -1,114 +1,113 @@
...
   if (input.size < 2 || input.speed < 1) {
-    return;
+    return
   }
 
-  resultWrapper.style.display = 'block';
+  resultWrapper.style.display = 'block'
 
-  var randomBits = Math.log(input.alphabet.length) / Math.LN2;
+  var randomBits = Math.log(input.uniqAlphabet.length) / Math.LN2
```
Comment (Kelin2025) [inline src/index.js]: Replaced `alphabet` to `uniqAlphabet`

File: src/index.js [hunk 4]
```diff
@@ -1,114 +1,113 @@
...
-  exampleElement.style.display = 'none';
+function example(input) {
+  exampleElement.style.display = 'none'
 
-  if (
-    input.alphabet.length < 2
-    || input.size <= 1
-  ) {
-    return;
+  if (input.uniqAlphabet.length < 2 || input.size <= 1) {
```
Comment (Kelin2025) [inline src/index.js]: Replaced `alphabet` to `uniqAlphabet`

File: src/index.js [hunk 5]
```diff
@@ -1,114 +1,113 @@
...
-  alphabetError.innerHTML = '';
-  speedError.innerHTML = '';
+function validate(input) {
+  sizeError.innerHTML = ''
+  alphabetError.innerHTML = ''
+  speedError.innerHTML = ''
 
-  if (input.alphabet.length < 2) {
-    alphabetError.innerHTML = 'The length of the alphabet should be 2 or more';
+  if (input.uniqAlphabet.length < 2) {
```
Comment (Kelin2025) [inline src/index.js]: Replaced `alphabet` to `uniqAlphabet`

File: src/index.js [hunk 6]
```diff
@@ -1,114 +1,113 @@
...
+    alphabetError.innerHTML =
+      'An alphabet should contain at least 2 <b>unique</b> symbols'
   }
 
-  if (input.alphabet.length > 256) {
-    alphabetError.innerHTML = 'An alphabet with than 256 symbols is not secure due to algorithm limitations';
+  if (input.uniqAlphabet.length > 256) {
```
Comment (Kelin2025) [inline src/index.js]: Replaced `alphabet` to `uniqAlphabet`

**Comment by alex7kom** [comment]
Thank you for the PR!

I've added eslint/prettier to the project. Could you please rebase your changes on the current master?

**Comment by alex7kom** [comment]
Actually I think it's better to error with a different message if alphabet contains duplicate symbols than simply remove them, so the user will know what happened.

**Comment by Kelin2025** [comment]
Added warning about duplicated symbols. I think we shouldn't remove duplicated symbols automatically because it may confuse people
image

File: src/index.js [hunk 7]
```diff
@@ -38,6 +38,9 @@ function calculate() {
   var input = {
     size: Number(sizeInput.value),
     alphabet: alphabetInput.value,
+    uniqAlphabet: alphabetInput.value.split('').reduce(function(res, letter) {
+      return res.includes(letter) ? res : res + letter;
```
Comment (alex7kom) [inline src/index.js]: `String#includes` is ES6, so it doesn't work in IE.

**Review by alex7kom — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 1 above)
Comment (Kelin2025) [reply src/index.js]: Changed .includes() to .indexOf()

**Comment by Kelin2025** [comment]
@Alex7Kom it's ok now

**Review by alex7kom — APPROVED** [review]

**Comment by alex7kom** [comment]
Thank you!

**Pull request merged**